README.md
# RanGen Fantasy Names
RanGen Fantasy Names is a Python application that procedurally generates names by joining a number of randomly generated syllables that are constructed following 'consonant-vowel-consonant' conventions. It uses a custom module (rangen_words) as the back-end for name generation, and a GUI front-end using PyQt5.

# About
This application works by running a function to generate a syllable that is created using consonant-vowel-consonant conventions. It may randomly omit the starting or ending consonant, and can generate a cluster of consonants such as "ck" or "th" intead of a single consonant, or it can generate a diphthong such as "ie" or "oo" instead of a single vowel. It will then string together a number of these syllables to form the generated name.

While using the GUI, the check boxes are used to select specific names you would like to work with. Selected names will also be ignored when re-rolling. Currently, you can copy selected names to your clipboard either as text or as an image. You can also export the selected names to a CSV file which will be appended after any existing entries. Many of the settings can be set by the user via inputs and saved to a config file. The 'About' menu in the application will give more information.

# Screenshots
[Screenshots](/screenshots)

![Alt text](/screenshots/v1-0-0_main.png?raw=true "Main Window")
![Alt text](/screenshots/v1-0-0_charsets.png?raw=true "Charset Window")

# Change Log
[Change Log](CHANGELOG.md)

# Requirements
[Requirements](requirements.txt)

# Other Files
JSON Config
-
The config is a JSON file and will save custom settings. The program will only generate this file if you ever choose to save your settings. You can save current settings in the program which will update the config file and will be the new settings that the program will reset to unless you choose to reset the config file back to system defaults. If the file is ever corrupted, it will always be able to use the default settings.

JSON Charset
-
The pre-packaged [charset file](/rangen/charsets.json) is just an example of how you can define your own character sets. The program will always be able to use a default character set if you don't have a charset file or the charset file is corrupted. Any custom character sets will need to contain each of the 5 dictionaries (vowels, diphthongs, consonants, beginning_clusters, and ending_clusters) to work correctly but the names (such as 'Orc') in the pre-packaged file can be whatever you choose, as well as the letters inside each dictionary.

CSV File
-
One of the current options is to export your names to a CSV file. In the short term, this is just so you can maintain a list of names you like as the names are appended to the file so it keeps track of all names you have exported in the past. In the long term, the program may do more with this file as more features are added.

# Road Map
The basic intended functions of the program have all been implemented. The next feature likely to be implemented will be the ability to create or modify character sets directly in the GUI and perhaps the ability to view exported names in the CSV file directly within the GUI.

# Credits
Lanecrest Tech © 2023

This program uses Qt Framework Essentials, specifically the PyQt5 binder from [Riverbank Computing](https://www.riverbankcomputing.com/)

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.


CHANGELOG.md
# Change Log

## v1.0.0 (04-02-23)
-Releases are now considered release candidate so beta numbering has been dropped.

-Settings have been moved back to the main display, but are now to the right of the name outputs instead of under. It was found that the program was easier to use when the names and settings panel were side by side.

-Made some minor aesthetic changes and updated About dialogue to reflect changes

-Now includes a charsets.json file to demonstrate how character sets should be constructed if adding your own or changing what is in there. The Default character set is still hardcoded into the program.

-Renamed main script to main.py and updated folder structure so the imported module and charsets file is in a subfolder

-Added open source license info for GPL v3

## v4.0.0b (03-23-23)
-Major update to rangen_words module, now supports custom character sets via a JSON file

-Major update to GUI, settings options are now accessed through a separate window. For now, you cannot access the settings window and the main window at the same time, but this will be a focus of the next major update

-The program will no longer generate/require either a settings config or a character set JSON file if the user chooses to only work with default settings. However, a charset json file is provided in the build to serve as an example of how to manipulate them correctly to add your own character sets

-All versions have had a 'b' appended to them to indicate that these are currently beta releases

## v3.3.1b (03-19-23)
-Changed some UI elements

-Updated About menu

## v3.3.0b (03-19-23)
-How long a name can be before it receives a split is now a user adjustable setting

-Letter generation probabilities can now be set to 0, but if all the cononants are set to 0, no names will generate, as the program would get stuck in an infinite loop as it tries to generate a name with at least one consonant and vowel.

-Added some more ending cluster pairings to the character set

-Made the menu bar easier to be read and changed some of the shortcuts

-Optimized the code for saving and resseting settings

## 3.2.0b (03-18-23)
-Replaced most of the buttons with a file menu system which includes shortcut keys

-When copying to clipboard, the selected names will now also be printed in the console so you see whats been copied before pasting

-Optimized parts of the code to make sweeping changes less painful in the future

-Settings can now be exported to a JSON config file and the program will use those settings if present. The config file will be automatically generated with system default settings if it doesn't exist, and you can reset your custom settings to the default settings if you wish. Most importantly, this allows you to keep settings you like so they will be saved between instances of running the program. If the config file is ever corrupt or missing values, it will try to rebuild parts or the entire of the file with the system defaults

-Errors with the config and the CSV file will now be printed in the console in addition to any outputs in the GUI

## 3.1.1b (03-16-23)
-Fixed a bug with slider values that caused the program to crash

-A couple of minor tweaks to presentation

-All new releases will be date stamped going forward. Exact times and dates of the 2.0 releases were not recorded but version 1.0 was relased on 3-9-23

## v3.1.0b (03-16-23)
-Added ability to choose the maximum number of syllables that can generate for each name (updated rangen_words to support this as a feature of the core module).

-As a side effect, for now there is no weight distribution to the number of syllables that will generate. This may be added back in the future.

-Fixed some references in code to other parts of code including commits that was missed in the 3.0 update

-Changed the UI a bit to account for the more user settings options

## v3.0.0b (03-15-23)
-Split the syllable/name generation functions into their own module. "rangen_words" is now its own script and contains the functions for word generation. this can allow it to be easily imported into other projects. rangen_fantasy_names is now just the GUI front end that uses rangen_words to make 'fantasy names.'

-Added ability for users to adjust many of the settings

-Removed some code in the name generation (rangen_words) that turned out to not be very useful

-Optimized some code and behaviors over all

## v2.3.0b
-Removed possibility for a letter to repeat itself more than twice in a row (maybe)

-Changed split names to have an apostrophe instead of a space.

-Updated pixmap spacing when copy as image

-Updated gui to better align widgets

## v2.2.5b
-Added an export function. This will export selected names to a .csv file and will append names if names already exist

-Fixed some typos and updated the About box

-Made four syllable names slightly less frequent

-There is now just one re-roll button and it ignores selected names. This is explained in the About box

## v2.2.0b
-Overhaul to the GUI layout

-Changed the name of the program and project

-Names now generate in their own text boxes and have corresponding check boxes to select them

-Can multi select names to copy to clipboard as either text or an image

-Can re-reroll names based on what is selected

-Added an 'About' dialog box and link to github project

## v2.1.0b
-Minor tweaks to the GUI

-Changed some of the values for syllable generation

-Added a few more diphthongs

## v2.0.0b
-Adjusted many of the conditions for generating syllables

-Now runs in PyQt5

-Can copy text out of the text box or copy as image for easy sharing

## v1.0.0b (03-09-23)
-Initial release, runs in console


fantasy_names.py
import random, os

vowels = ['a', 'e', 'i', 'o', 'u', 'y']
diphthongs = ['ae', 'ai', 'au', 'ea', 'ee', 'ei', 'eo', 'eu', 'ia', 'ie', 'io', 'oa', 'oi', 'oo', 'ou']
consonants = ['b', 'c', 'd', 'f', 'g', 'h', 'j', 'k', 'l', 'm', 'n', 'p', 'q', 'r', 's', 't', 'v', 'w', 'x', 'z']
beginning_clusters = ['bl', 'br', 'ch', 'cl', 'cr', 'dr', 'fl', 'fr', 'gl', 'gr', 'gw', 'kl', 'ph', 'pl', 'pr', 'qu', 'rh', 'sc', 'sh', 'sk', 'sl', 'sm', 'sn', 'sp', 'st', 'sw', 'th', 'tr', 'tw', 'wh', 'wr', 'zh', 'scr', 'shr', 'sph', 'spl', 'spr', 'squ', 'str', 'thr']
ending_clusters = ['ch', 'ck', 'ft', 'ld', 'lt', 'nd', 'ng', 'nk', 'nt', 'ph', 'pt', 'rd', 'rk', 'sh', 'sk', 'sp', 'st', 'th', 'dd', 'ff', 'gg', 'll', 'mm', 'pp', 'rr', 'ss', 'tt', 'zz', 'rth', 'tch']

def generate_syllable():
    syllable = ''
    prev_char = ''
    # generate the first part of the syllable
    if random.random() < 0.15:
        # generate a beginning cluster
        beginning_cluster = random.choice(beginning_clusters)
        if len(syllable) > 0 and prev_char == syllable[-1]:
            beginning_cluster = random.choice([b for b in beginning_clusters if b[0] != prev_char])
        syllable += beginning_cluster
        prev_char = beginning_cluster[-1]
    else:
        # generate a consonant or no consonant
        if random.random() < 0.8:
            consonant = random.choice(consonants)
            if len(syllable) > 0 and prev_char == syllable[-1]:
                consonant = random.choice([c for c in consonants if c != prev_char])
            syllable += consonant
            prev_char = consonant
    # generate the middle part of the syllable
    if random.random() < 0.7:
        # genearate a vowel
        vowel = random.choice(vowels)
        if len(syllable) > 0 and prev_char == syllable[-1]:
            vowel = random.choice([v for v in vowels if v != prev_char])
        syllable += vowel
        prev_char = vowel
    else:
        # generate a diphthong
        diphthong = random.choice(diphthongs)
        if len(syllable) > 0 and prev_char == syllable[-1]:
            diphthong = random.choice([d for d in diphthongs if d[0] != prev_char])
        syllable += diphthong
        prev_char = diphthong[-1]
    # generate the last part of the syllable
    if random.random() < 0.15:
        # generate an ending cluster
        ending_cluster = random.choice(ending_clusters)
        if len(syllable) > 0 and prev_char == syllable[-1]:
            ending_cluster = random.choice([e for e in ending_clusters if e[0] != prev_char])
        syllable += ending_cluster
        prev_char = ending_cluster[-1]
    else:
        # generate a consonant or no consonant
        if random.random() < 0.5:
            consonant = random.choice(consonants)
            if len(syllable) > 0 and prev_char == syllable[-1]:
                consonant = random.choice([c for c in consonants if c != prev_char])
            syllable += consonant
            prev_char = consonant
    return syllable

def generate_name():
    name = ''
    # keep generating a name until it has at least one consonant or cluster
    while not any(c in consonants or c in beginning_clusters or c in ending_clusters for c in name):
        num_syllables = random.randint(1, 4)
        syllables = [generate_syllable() for _ in range(num_syllables)]
        name = ''.join(syllables)
        # insert a space between two syllables if there are more than a certain number of total letters generated
        if len(''.join(syllables)) > 10:
            i = random.randint(1, len(syllables) - 1)
            syllables.insert(i, ' ')
        name = ''.join(syllables)
    return name.title()

while True:
    print("Here are 10 randomly generated fantasy names:\n")
    
    for i in range(10):
        print(generate_name())
        
    user_input = input("\nPress 'Enter' to generate a new list or type 'n' to quit: ")
    if user_input.lower() == 'n':
        break
    os.system('cls' if os.name == 'nt' else 'clear')


fantasy_name_generator.py
import random, os
from PyQt5 import QtWidgets, QtGui, QtCore
from PyQt5.QtWidgets import QApplication, QWidget, QVBoxLayout, QHBoxLayout, QLabel, QPushButton, QTextEdit, QStyleFactory, QMessageBox
from PyQt5.QtGui import QFont, QPalette, QColor, QDesktopServices, QPixmap, QClipboard
from PyQt5.QtCore import Qt, QUrl

class FantasyNameGenerator(QWidget):
    # initialize the app and set some variables
    def __init__(self):
        super().__init__()
        # define the letters
        self.vowels = ['a', 'e', 'i', 'o', 'u', 'y']
        self.diphthongs = ['ae', 'ai', 'au', 'ay', 'ea', 'ee', 'ei', 'eo', 'eu', 'ey', 'ia', 'ie', 'io', 'oa', 'oi', 'oo', 'ou', 'oy']
        self.consonants = ['b', 'c', 'd', 'f', 'g', 'h', 'j', 'k', 'l', 'm', 'n', 'p', 'q', 'r', 's', 't', 'v', 'w', 'x', 'z']
        self.beginning_clusters = ['bl', 'br', 'ch', 'cl', 'cr', 'dr', 'fl', 'fr', 'gl', 'gr', 'gw', 'kl', 'ph', 'pl', 'pr', 'qu', 'rh', 'sc', 'sh', 'sk', 'sl', 'sm', 'sn', 'sp', 'st', 'sw', 'th', 'tr', 'tw', 'wh', 'wr', 'zh', 'scr', 'shr', 'sph', 'spl', 'spr', 'squ', 'str', 'thr']
        self.ending_clusters = ['ch', 'ck', 'ft', 'ld', 'lt', 'nd', 'ng', 'nk', 'nt', 'ph', 'pt', 'rd', 'rk', 'sh', 'sk', 'sp', 'st', 'th', 'dd', 'ff', 'gg', 'll', 'mm', 'pp', 'rr', 'ss', 'tt', 'zz', 'rth', 'tch']

        self.init_ui()
        self.generate_names()
        
    # create a function to copy to clipboard as an image
    def copy_to_clipboard(self):
        pixmap = self.names_text_edit.grab()
        clipboard = QApplication.clipboard()
        clipboard.setPixmap(pixmap)

    # set and organize the widgets
    def init_ui(self):
        # create a function to open links
        def handle_link(url):
            QDesktopServices.openUrl(QUrl(url))
        
        # define buttons and what they do
        self.generate_button = QPushButton('Generate')
        self.generate_button.setFixedWidth(125)
        self.generate_button.clicked.connect(self.generate_names)
        self.copy_button = QPushButton('Copy as Image')
        self.copy_button.setFixedWidth(125)
        self.copy_button.clicked.connect(self.copy_to_clipboard)
        self.quit_button = QPushButton('Quit')
        self.quit_button.setFixedWidth(125)
        self.quit_button.clicked.connect(QApplication.instance().quit)

        # define the text edit box that will contain the generated names
        self.names_text_edit = QTextEdit()
        text_edit_font = QFont("Consolas", 12)
        text_edit_font.setBold(True)
        self.names_text_edit.setFont(text_edit_font)
        self.names_text_edit.setReadOnly(True)

        # create the buttons in a horizontal box
        button_layout = QHBoxLayout()
        button_layout.addWidget(self.generate_button)
        button_layout.addWidget(self.copy_button)
        button_layout.addWidget(self.quit_button)
 
        # create the footer in a horizontal box
        foot_layout= QHBoxLayout()
        foot_link = QLabel('Fantasy Name Generator ©2023 <a href="https://www.lanecrest.com/">Lanecrest Tech</a>')
        foot_link.setOpenExternalLinks(True)
        foot_link.linkActivated.connect(handle_link)
        foot_layout.addWidget(foot_link, alignment=QtCore.Qt.AlignHCenter)

        # create a veritcal box for the main display and layer the buttons, text edit, and footer in that order
        main_layout = QVBoxLayout()
        main_layout.addLayout(button_layout)
        main_layout.addWidget(QLabel('Click \'Generate\' for another set of random names'))
        main_layout.addWidget(self.names_text_edit)
        main_layout.addLayout(foot_layout)
        
        # finalize the window display and set the window title
        self.setLayout(main_layout)
        self.setWindowTitle('Fantasy Name Generator 2.1')

    # define how syllables are generated
    def generate_syllable(self, prev_char=''):
        syllable = ''
        # generate the first part of the syllable
        if random.random() < 0.2:
            # generate a beginning cluster
            beginning_cluster = random.choice(self.beginning_clusters)
            if len(syllable) > 0 and prev_char == syllable[-1]:
                beginning_cluster = random.choice([b for b in self.beginning_clusters if b[0] != prev_char])
            syllable += beginning_cluster
            prev_char = beginning_cluster[-1]
        else:
            # generate a consonant or no consonant
            if random.random() < 0.8:
                consonant = random.choice(self.consonants)
                if len(syllable) > 0 and prev_char == syllable[-1]:
                    consonant = random.choice([c for c in self.consonants if c != prev_char])
                syllable += consonant
                prev_char = consonant
        # generate the middle part of the syllable
        if random.random() < 0.75:
            # genearate a vowel
            vowel = random.choice(self.vowels)
            if len(syllable) > 0 and prev_char == syllable[-1]:
                vowel = random.choice([v for v in self.vowels if v != prev_char])
            syllable += vowel
            prev_char = vowel
        else:
            # generate a diphthong
            diphthong = random.choice(self.diphthongs)
            if len(syllable) > 0 and prev_char == syllable[-1]:
                diphthong = random.choice([d for d in self.diphthongs if d[0] != prev_char])
            syllable += diphthong
            prev_char = diphthong[-1]
        # generate the last part of the syllable
        if random.random() < 0.15:
            # generate an ending cluster
            ending_cluster = random.choice(self.ending_clusters)
            if len(syllable) > 0 and prev_char == syllable[-1]:
                ending_cluster = random.choice([e for e in self.ending_clusters if e[0] != prev_char])
            syllable += ending_cluster
            prev_char = ending_cluster[-1]
        else:
            # generate a consonant or no consonant
            if random.random() < 0.6:
                consonant = random.choice(self.consonants)
                if len(syllable) > 0 and prev_char == syllable[-1]:
                    consonant = random.choice([c for c in self.consonants if c != prev_char])
                syllable += consonant
                prev_char = consonant
        return syllable

    # generate a name based on the syllable generation rules
    def generate_name(self):
        name = ''
        # keep generating a name until it has at least one consonant or cluster
        while not any(c in self.consonants or c in self.beginning_clusters or c in self.ending_clusters for c in name):
            num_syllables = random.choices([1, 2, 3, 4], weights=[1, 3, 3, 3])[0]
            syllables = [self.generate_syllable() for _ in range(num_syllables)]
            name = ''.join(syllables)
            # insert a space between two syllables if there are more than a certain number of total letters generated
            if len(''.join(syllables)) > 10:
                i = random.randint(1, len(syllables) - 1)
                syllables.insert(i, ' ')
            name = ''.join(syllables)
        return name.title()

    # a function that when called generates multiple names at a time
    def generate_names(self):
        self.names_text_edit.clear()
        for i in range(8):
            self.names_text_edit.append(self.generate_name() + '\n')
        
# load the app. in an if statement so it doesn't break if called in another app
if __name__ == '__main__':
    # set the palette colors
    palette = QPalette()
    palette.setColor(QPalette.Window, QColor('#d5a05a')) # window background color
    palette.setColor(QPalette.WindowText, QColor('#ffefe0')) # window font color
    palette.setColor(QPalette.Base, QColor('#704214')) # text edit background color
    palette.setColor(QPalette.Text, QColor('#ffefe0')) # text edit font color
    palette.setColor(QPalette.Button, QColor('#704214')) # button background color
    palette.setColor(QPalette.ButtonText, QColor('#ffefe0')) # button font color
    palette.setColor(QPalette.AlternateBase, QColor('#b8803c')) # table and dropdown background color
    palette.setColor(QPalette.ToolTipBase, QColor('#b8803c')) # tooltip background color
    palette.setColor(QPalette.ToolTipText, QColor('#ffefe0')) # tooltip font color
    palette.setColor(QPalette.BrightText, QColor('#ffefe0')) # highlighted font color
    
    # load the app
    app = QApplication([])
    app.setStyle(QStyleFactory.create('Fusion'))
    app.setPalette(palette)
    fantasy_name_generator = FantasyNameGenerator()
    fantasy_name_generator.setGeometry(100, 100, 800, 600)
    fantasy_name_generator.show()
    app.exec_()


rangen/rangen_words.py
import random, json, os

# class object to use for easy character set indexing
class CharSet:
    pass

# function to load the character set
def load_charset(load_set=None):
    charset = CharSet()
    charset_dir = os.path.dirname(os.path.realpath(__file__))   # make sure charsets file is loaded correctly if the script is imported
    charset_file = os.path.join(charset_dir, 'charsets.json')
    default_set = {
        # if building your own character sets for a json file, each of these dictionaries need to be wrapped in, for example: 'CharSetName': { }
        'vowels': ['a', 'e', 'i', 'o', 'u', 'y'],
        'diphthongs': ['ae', 'ai', 'au', 'ay', 'ea', 'ee', 'ei', 'eo', 'eu', 'ey', 'ia', 'ie', 'io', 'oa', 'oe', 'oi', 'oo', 'ou', 'oy', 'ua', 'ue', 'ui'],
        'consonants': ['b', 'c', 'd', 'f', 'g', 'h', 'j', 'k', 'l', 'm', 'n', 'p', 'q', 'r', 's', 't', 'v', 'w', 'x', 'z'],
        'beginning_clusters': ['bl', 'br', 'ch', 'cl', 'cr', 'dr', 'fl', 'fr', 'gl', 'gr', 'gw', 'kl', 'ph', 'pl', 'pr', 'qu', 'rh', 'sc', 'sh', 'sk', 'sl', 'sm', 'sn', 'sp', 'st', 'sw', 'th', 'tr', 'tw', 'wh', 'wr', 'zh', 'scr', 'shr', 'sph', 'spl', 'spr', 'squ', 'str', 'thr'],
        'ending_clusters': ['ch', 'ck', 'ct', 'dd', 'ff', 'ft', 'gg', 'ld', 'lf', 'lk', 'll', 'lm', 'lt', 'mm', 'mp', 'nd', 'ng', 'nk', 'nt', 'ph', 'pp', 'pt', 'rd', 'rk', 'rr', 'sh', 'sk', 'sp', 'ss', 'st', 'th', 'tt', 'zz', 'rth', 'tch']
    }
    # load json file with custom character sets if it exists
    if os.path.exists(charset_file):
        try:
            with open(charset_file, 'r') as f:
                char_file = json.load(f)
                # load the character set that is passed through load_set
                if load_set in char_file:
                    char_set = char_file[load_set]
                # load the default character set if the json file exists but None is still passed in load_set
                elif load_set == None:
                    char_set = default_set  
                # error handling to load the default character set if the charset file exists but the value passed through load_set doesn't exist
                else:
                    print(f'Error: Could not find a character set named "{load_set}" in the charset file. Default character set will be loaded.')
                    char_set = default_set
        # error handling to load the default character set if the json file is corrupt
        except json.JSONDecodeError:
            print('Error: The charset file is corrupt. Default character set will be loaded.')
            char_set = default_set
    # load the default character set if there is no json file
    else:
        char_set = default_set    
    for key, value in char_set.items():
        setattr(charset, key, value)
    return charset

# function to define how syllables are generated
def rangen_syllable(load_set=None, beg_cons_prob=0, beg_cluster_prob=0, vowel_prob=0, end_cons_prob=0, end_cluster_prob=0):
    char_set = load_charset(load_set)
    syllable = ''
    # generate the first part of the syllable
    if random.random() < beg_cons_prob: # if set to 1, a consonant will always generate instead of a cluster or nothing
        consonant = random.choice(char_set.consonants)
        syllable += consonant
    else:
        if random.random() < beg_cluster_prob:  # if set to 1, a cluster will always generate instead of nothing
            beginning_cluster = random.choice(char_set.beginning_clusters)
            syllable += beginning_cluster
    # generate the middle part of the syllable
    if random.random() < vowel_prob:    # if set to 1, a vowel will always generate instead of a diphthong
        vowel = random.choice(char_set.vowels)
        syllable += vowel
    else:
        diphthong = random.choice(char_set.diphthongs)
        syllable += diphthong
    # generate the last part of the syllable
    if random.random() < end_cons_prob: # if set to 1, a consonant will always generate instead of a cluster or nothing
        consonant = random.choice(char_set.consonants)
        syllable += consonant
    else:
        if random.random() < end_cluster_prob:  # if set to 1, a cluster will always generate instead of nothing
            ending_cluster = random.choice(char_set.ending_clusters)
            syllable += ending_cluster
    return syllable

# function to generate a word based on the syllable generation function
def rangen_word(load_set=None, beg_cons_prob=0, beg_cluster_prob=0, vowel_prob=0, end_cons_prob=0, end_cluster_prob=0, max_syllables=0, word_split=False, split_char='', word_long=0):
    char_set = load_charset(load_set)
    word = ''
    while True:
        num_syllables = random.randint(1, int(max_syllables))
        syllables = [rangen_syllable(load_set, beg_cons_prob, beg_cluster_prob, vowel_prob, end_cons_prob, end_cluster_prob) for _ in range(num_syllables)]
        word = ''.join(syllables)
        # check if a single letter repeats itself too many times and regenerate the word if so
        for i, letter in enumerate(word):
            if i > 1 and letter == word[i-1] and letter == word[i-2]:
                word = ''
                break
        # check if all the consonant probabilities were set to 0 which would cause an infinite loop, so instead it just prints empty words
        if beg_cons_prob == 0 and beg_cluster_prob == 0 and end_cons_prob == 0 and end_cluster_prob == 0:
            word = ''
            break
        # check to only print words that contain at least one consonant. if there are no consonants then an empty word will print
        if any(c in char_set.consonants or c in char_set.beginning_clusters or c in char_set.ending_clusters for c in word):
            break
    # insert a character between syllables if there are more than a certain number of total letters generated
    if len(word) >= word_long and word_split and len(syllables) > 1:
        i = random.randint(1, len(syllables) - 1)
        syllables.insert(i, split_char)
        word = ''.join(syllables)
    return word

# debug testing since the script is intended to be imported into other projects. using some preset values and default character set.
if __name__ == "__main__":
    word = rangen_word(load_set=None, beg_cons_prob=0.8, beg_cluster_prob=0.2, vowel_prob=0.75, end_cons_prob=0.6, end_cluster_prob=0.15, max_syllables=4, word_split=False, split_char='', word_long=10)
    print(word)